### PR TITLE
Add markdown converted notebook

### PR DESCRIPTION
**Pull request by JonnySpruce**

### CODE REVIEW

File: diffedit.md [hunk 1]
```diff
@@ -0,0 +1,752 @@
+# DiffEdit
+
+In this blog post, we will be demonstrating how to use the DiffEdit technique described in [this paper](https://arxiv.org/abs/2210.11427) (TODO: should this be more of a citation format?), to use a diffusion model to modify just one part of an existing image using simple text prompts. DiffEdit utilises the diffusion model which is used to predict where noise is in an image, typically as a way of generating images using text prompts. In this blog post we assume you have basic knowledge of Stable Diffusion and how it works, as this is the diffusion model that we will be using throughout our examples. If you want to read more about diffusion models, then check out our other blog post about it here [TODO]().
```
Comment (chrisprice) [inline diffedit.md]: I think your existing referencing style is fine for a blog post.

Comment (chrisprice) [inline diffedit.md]: For the blog you'll also need an exec summary to be used on the contents page. If you don't specify one you'll end up with one autogenerated from the first n characters which is normally guff.

File: diffedit.md [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+The DiffEdit paper proposes a new way to use the noise prediction which a diffusion model provides to us to automatically create a mask which can then be used to limit the area where an image can be changed. In the sections below we will first work through automatically generating that mask and a few techniques for improving that process, and then demonstrate using that mask to only edit one part of our original image.
+
+It does this in two main steps:
+
+1. Generate a mask over the original image
+1. Use this mask to only modify the original image based on the new prompt
+
+We will work through these steps in the sections below.
+
+**TODO: also show example image here?**
```
Comment (chrisprice) [inline diffedit.md]: You might have already meant this but I would suggest a pipeline overview type image

File: diffedit.md [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+
+We will work through an example where we use the DiffEdit technique to modify an image of a horse into an image of a zebra.
+
+Many diffusion models already can be used in an image to image pipeline which can take in an original image and tweak it based on a new prompt, which may at first seem similar to the DiffEdit technique. Image to image is fairly limited however, as all parts of the image will be changed by this process, you cannot limit the modifications of the image to a specific area.
```
Comment (chrisprice) [inline diffedit.md]: There are some pretty dense sentences in the next couple of paragraphs. Might be worth a revisit to see if you can simplify them.

File: diffedit.md [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+    
+![png](diffedit_files/diffedit_18_0.png)
+    
+
+
+
+Getting the difference in noise prediction between the two prompts is achieved by adding 50% noise to our original image, and then giving our diffusion model the two prompts and ask it to predict noise for both `"horse"` and `"zebra"`.
+We can then take the difference in the noise predictions to create our mask.
+
+This works because the original image is already an image of a horse, and so our `"horse"` prompt should create a noise prediction which is evenly distributed. For the `"zebra"` prompt however, while the background is plausible given the prompt, the horse in the foreground isn't very zebra-like at all, and so our model will predict that there must be additional noise in this area. This means that if we take the difference between the two predictions (averaged out over 10 times), then we should get a pretty good outline of a horse.
```
Comment (chrisprice) [inline diffedit.md:38]: Great explanation. Love it!

File: diffedit.md [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+```
+
+
+
+
+    torch.Size([1, 4, 64, 64])
+
+
+
+We can see that the image has been converted into a 1x64x64x4 tensor, which is the latent format that our particular diffusion model can work with. The images below show you what those latents look like over the 4 layers:
```
Comment (chrisprice) [inline diffedit.md:54]: You might want to head off the obvious question here of whether these layers map onto a more common representation like RGBA. I assume the answer is no...

File: diffedit.md [hunk 6]
```diff
@@ -0,0 +1,752 @@
...
+
+## Prompt to Text Embeddings
+
+We have now converted our original image into the correct format for our model, and we also need to do this for the text prompts we will be providing.
+
+To do this we first use a tokenizer to map the text into number values, and then we use an encoder to convert them into tensors in a format that the model can work with. We have stored this behaviour in a handy function so that we can reuse this logic again later for other prompts.
+
+
+```python
+def create_text_embeddings(prompts, batch_size=1):
```
Comment (chrisprice) [inline diffedit.md:71]: N.B. I know next to nothing about this area.

Is it beneficial to `to('cuda')` these relatively small tensors for a relatively small number of relatively simple operations?

Comment (chrisprice) [reply diffedit.md:71]: As opposed to `to('cuda')`-ing the result?

File: diffedit.md [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+## Noising the original image
+
+To add noise to our original image to the correct level, we will utilise the scheduler. The scheduler typically manages noise throughout the image creation process for diffusion models, and it can also be used to manually add a specific level of noise to an image.
+
+First we need to set up the scheduler, and provide it with the configuration values that our diffusion model expects. We have used a DDIM scheduler, which is the type of scheduler that is recommended by the DiffEdit paper as it allows for more consistent results between the original image and the edited image.
+
+We'll define this setup as a function as we'll be using it again later when we generate the DiffEdit image. The argument `num_inference_steps` is the total number of steps that our model would take to fully denoise an image. With the version of Stable Diffusion and the scheduler we have chosen, 50 steps usually works well.
+
+
+```python
```
Comment (chrisprice) [inline diffedit.md:131]: A general comment is that while the notebook format encourages lots of code cells, typically in blog posts there's more prose than code. I would be tempted to trim some of the code blocks (maybe 50%).

File: diffedit.md [hunk 8]
```diff
@@ -0,0 +1,752 @@
...
+We can then take a look at our noised latents to get an idea of what it they look like:
+
+
+```python
+latents_to_grid_bw(noisy_latents.squeeze().detach().cpu(), 4)
+```
+
+
+    
+![png](diffedit_files/diffedit_43_0.png)
```
Comment (chrisprice) [inline diffedit.md]: All of these and the subsequent images do a great job of visualising your progress.

File: diffedit.md [hunk 9]
```diff
@@ -0,0 +1,752 @@
...
+callback = (lambda latents: itermediate_latents.append(latents))
+
+# create text embeddings for new_prompt "zebra"
+text_embeddings = create_text_embeddings([new_prompt])
+
+mask = torch.from_numpy(mask_np).type(torch.half).to("cuda")
+```
+
+We use an image to image pipeline to pass in our original image, since we still want our zebra to have a similar pose to the horse which is present in our original image, so we add a lot of noise to the original image (~75%) and then we will use this as the starting point for our image generation of our zebra. We can see what that initial noise looks like below:
+TODO: should we show the latents or the decoded image?
```
Comment (chrisprice) [inline diffedit.md]: I'd argue neither. They both look like a poorly tuned analogue TV!

File: diffedit.md [hunk 10]
```diff
@@ -0,0 +1,752 @@
...
+images_to_grid([*i, to_image(final_latents)], 3, titles=[*labels, f"step {num_inference_steps}"])
+```
+
+
+    
+![png](diffedit_files/diffedit_86_0.png)
+    
+
+
+## Putting it all together in functions
```
Comment (chrisprice) [inline diffedit.md]: As this is already a rather long (but thoroughly enjoyable!) post. I'd recommend dropping this section. You can always link to a repo with the full code in it.

File: diffedit.md [hunk 11]
```diff
@@ -0,0 +1,752 @@
...
+    
+![png](diffedit_files/diffedit_92_2.png)
+    
+
+
+We can see a very nice mask generation here, along with some slightly wonky looking pears, that's mainly an issue with our diffusion model's image generation in this case rather than anything that we've done specifically, so we'll accept slightly wonky pears for now.
+
+## Limitations
+
+Something else which is worth considering is the limitations of the DiffEdit mask creation technique. If there are huge differences between the original image caption and the query string then the mask that is generated is not as accurate, which has the knock-on effect of leading to poor final image generation. Take our example below, we'll use the same original image and caption as before, but this time use the string `"bowl of cars"` as the query. As we can see the mask creation includes random areas of the background due to the lack of similarities with the original image and the intended query, and generates a rather odd final image.
```
Comment (chrisprice) [inline diffedit.md]: n.b. if you drop the previous section, you might need to rework the start of this paragraph.

File: diffedit.md [hunk 12]
```diff
@@ -0,0 +1,752 @@
...
+    
+![png](diffedit_files/diffedit_95_1.png)
+    
+
+
+
+    
+![png](diffedit_files/diffedit_95_2.png)
+    
+
```
Comment (chrisprice) [inline diffedit.md:409]: You need a conclusion to ease the reader's landing back into the real world. Random ideas for a conclusion, pick and choose as you see fit -
We set out to do this, it worked as/not as we expected. It was more/less interesting/challenging/fun that expected. If I did it again I would do/not do.... In doing this work I came across this other area which I'm interested in investigating.

**Review by chrisprice — COMMENTED** [review]
Really enjoyed the post. Great work! Just a few comments to bring in more in line stylistically with the other blog posts and a few responses to the TODOs.

File: diffedit.md (hunk 9 above)
Comment (JonnySpruce) [reply diffedit.md]: You're right, we've already shown noisy latents before so we probably don't need to show them again.

File: diffedit.md (hunk 6 above)
Comment (JonnySpruce) [reply diffedit.md:71]: I'll be honest, I don't know much about the encoding process, but it seems like at least one of the functions we're calling expects the tensor we pass in to already be on the GPU as I get an error about it being on an unexpected device (cpu) if I edit it to only move to the GPU at the end.

File: diffedit.md [hunk 13]
```diff
@@ -358,41 +336,17 @@ all_preds.min(), all_preds.max()
...
-```python
-possible_thresholds = [0.01, 0.1, 0.2, 0.3, 0.5]
-masks = [(all_preds > threshold) for threshold in possible_thresholds]
-masks_to_grid(original_image, masks, len(possible_thresholds), titles=possible_thresholds)
-```
-
-
-    
-![png](diffedit_files/diffedit_64_0.png)
+We can write try a few different threshold values and visualise the resulting mask. This will allow us to choose a value that works well for our image:
```
Comment (jstrong-scottlogic) [inline diffedit.md]: write try -> try

File: diffedit.md [hunk 14]
```diff
@@ -552,201 +460,52 @@ for i, t in enumerate(tqdm(denoising_scheduler.timesteps[start_step:])):
...
-
-```python
-caption = "bowl of fruits"
-query = "bowl of cars"
-
-diffedit(fruit_bowl_image, caption, query, noise_strength=1)
-```
+## Conclusion
 
+When working on this DiffEdit pipeline we set out to implement the techniques in the paper using our own custom pipeline. As you can see in the example above, we were able to acheive this with accurate mask generation and decent final image replacement. This shows that DiffEdit is a very powerful technique, and it was suprisingly easy to adapt an img2img pipeline to work with the DiffEdit technique. The ability to pass in just two different text prompts and have an accurate mask generated as a result was truly impressive.
```
Comment (jstrong-scottlogic) [inline diffedit.md]: *achieve, *surprisingly

File: diffedit.md [hunk 15]
```diff
@@ -552,201 +460,52 @@ for i, t in enumerate(tqdm(denoising_scheduler.timesteps[start_step:])):
...
 
-      0%|          | 0/50 [00:00<?, ?it/s]
+There is still more for us to explore with this technique, from tweaking the various parameters further to tune both the mask creation and the image generation, to further understanding what prompts give the best results. It would also be interesting to adapt this to work with Stable Diffusion 2 or SDXL and see how that affects both the mask creation and final image generation.
 
+TODO: maybe talk about google magic eraser?
```
Comment (jstrong-scottlogic) [inline diffedit.md]: Would be interesting though I think mention of SDXL and Stable Diffusion 2 may be sufficient.

File: diffedit.md [hunk 16]
```diff
@@ -1,21 +1,23 @@
...
+In this blog post, we will be demonstrating how to use the DiffEdit technique described in [this paper](https://arxiv.org/abs/2210.11427), to use a diffusion model to modify just one part of an existing image using simple text prompts. DiffEdit utilises the diffusion model which is used to predict where noise is in an image, typically as a way of generating images using text prompts. In this blog post we assume you have basic knowledge of Stable Diffusion and how it works, as this is the diffusion model that we will be using throughout our examples. If you want to read more about diffusion models, then check out our other blog post about it here [TODO]().
```
Comment (jstrong-scottlogic) [inline diffedit.md]: I cannot see another SL blog post on diffusion unfortunately :-(

File: diffedit.md (hunk 5 above)
Comment (JonnySpruce) [reply diffedit.md:54]: You're right, let me know if the new added sentence makes sense.

File: diffedit.md (hunk 7 above)
Comment (JonnySpruce) [reply diffedit.md:131]: I've removed all of the blocks which involve just displaying the images, is that enough or should we also be removing other sections? One area that possibly could be removed as it pretty much exactly matches the standard image generation way of doing things is the section on text encoding.

File: diffedit.md (hunk 16 above)
Comment (JonnySpruce) [reply diffedit.md]: It was in my head that at some point someone in our study group would write one. I might try and do it myself next week. Not sure if we can edit the post to re-add this point at some point in the future if another blog is added that we can link to?

File: diffedit.md (hunk 3 above)
Comment (JonnySpruce) [reply diffedit.md]: I've reworked this section, let me know if it looks okay.

File: diffedit.md (hunk 12 above)
Comment (JonnySpruce) [reply diffedit.md:409]: I've now added a conclusion. Let me know if that's enough or if there's anything else you think we should add

File: diffedit.md (hunk 3 above)
Comment (chrisprice) [reply diffedit.md]: It's better but maybe could be improved a bit further. n.b. I'm only labouring it because it's the start of the post where you need to build trust with the reader that you're not going to leave them baffled. 

Here's GPT4's suggestion -

> In this post, we'll walk through an example of using the DiffEdit technique to transform an image of a horse into a zebra.
> 
> Many existing diffusion models can be utilized in an image-to-image pipeline. This process takes an original image and modifies it based on a new prompt, which may seem similar to the DiffEdit technique at first glance. However, the image-to-image method has its limitations. For instance, it alters all parts of the image, and there's no way to restrict the modifications to a specific area.
> 
> The DiffEdit paper introduces a novel approach. It suggests using the noise prediction provided by a diffusion model to automatically generate a mask. > This mask can then be used to confine the area of the image that can be altered. 
> 
> In the following sections, we'll first explore how to automatically generate this mask and discuss a few techniques to enhance this process. Subsequently, we'll demonstrate how to use this mask to edit only a specific part of our original image.

Comment (chrisprice) [reply diffedit.md]: Scratch that I was reviewing the old version.

File: diffedit.md (hunk 5 above)
Comment (chrisprice) [reply diffedit.md:54]: LGTM

File: diffedit.md (hunk 12 above)
Comment (chrisprice) [reply diffedit.md:409]: LGTM

File: diffedit.md (hunk 6 above)
Comment (chrisprice) [reply diffedit.md:71]: 🤷

File: diffedit.md (hunk 7 above)
Comment (chrisprice) [reply diffedit.md:131]: I still think there's a few too many. There's quite a few places where you initialise/assign variables. In most cases the variable name is sufficient to follow the explanation in the text. The specific values would only be required if you were trying to reproduce the post and the link to the repo should cover off that use case.

File: diffedit.md [hunk 17]
```diff
@@ -258,17 +200,12 @@ def predict_noise(scheduler, latents, combined_embeddings, guidance_scale, times
     return pred_original, pred_new
 ```
 
-The paper states that repeating the noise prediction 10 times allowed for a good mask to be generated, so we will use that value here.
+The paper states that repeating the noise prediction 10 times allowed for a good mask to be generated, so we will use that value here. We will then loop through and make our noise predictions. Each loop we set a different seed so that that added noise will be different, and then we take the difference in the noise predictions for our two prompts, and then store this difference in our variable `all_preds_base`. At the end of all the loops we divide `all_preds_base` by our number of loops to get an average difference over our 10 predictions.
```
Comment (jstrong-scottlogic) [inline diffedit.md]: "that that added noise" -> "that the added noise" ? Not certain
